CMake/Findmlpack.cmake

```diff
@@ -0,0 +1,220 @@
+#  ENSMALLEN_INCLUDE_DIR - list of required include directories
+#  ENSMALLEN_VERSION_MAJOR - major version number
+#  ENSMALLEN_VERSION_MINOR - minor version number
+#  ENSMALLEN_VERSION_PATCH - patch version number
+#  ENSMALLEN_VERSION_STRING - version number as a string (ex: "1.0.4")
+#  ENSMALLEN_VERSION_NAME - name of the version (ex: "Antipodean Antileech")
+
+file(GLOB ENSMALLEN_SEARCH_PATHS
+    ${CMAKE_BINARY_DIR}/deps/ensmallen-[0-9]*.[0-9]*.[0-9]*)
+find_path(ENSMALLEN_INCLUDE_DIR
+  NAMES ensmallen.hpp
+  PATHS ${ENSMALLEN_SEARCH_PATHS}/include)
+
+if(ENSMALLEN_INCLUDE_DIR)
+  # ------------------------------------------------------------------------
+  #  Extract version information from <ensmallen>
+  # ------------------------------------------------------------------------
+
+  set(ENSMALLEN_VERSION_MAJOR 0)
+  set(ENSMALLEN_VERSION_MINOR 0)
+  set(ENSMALLEN_VERSION_PATCH 0)
+  set(ENSMALLEN_VERSION_NAME "unknown")
+
+  if(EXISTS "${ENSMALLEN_INCLUDE_DIR}/ensmallen_bits/ens_version.hpp")
+
+    set(ENSMALLEN_FOUND YES)
+
+    # Read and parse armdillo version header file for version number
+    file(READ "${ENSMALLEN_INCLUDE_DIR}/ensmallen_bits/ens_version.hpp"
+        _ensmallen_HEADER_CONTENTS)
+    string(REGEX REPLACE ".*#define ENS_VERSION_MAJOR ([0-9]+).*" "\\1"
+        ENSMALLEN_VERSION_MAJOR "${_ensmallen_HEADER_CONTENTS}")
+    string(REGEX REPLACE ".*#define ENS_VERSION_MINOR ([0-9]+).*" "\\1"
+        ENSMALLEN_VERSION_MINOR "${_ensmallen_HEADER_CONTENTS}")
+    string(REGEX REPLACE ".*#define ENS_VERSION_PATCH ([0-9]+).*" "\\1"
+        ENSMALLEN_VERSION_PATCH "${_ensmallen_HEADER_CONTENTS}")
+
+    # WARNING: The number of spaces before the version name is not one.
+    string(REGEX REPLACE
+        ".*#define ENS_VERSION_NAME\ +\"([0-9a-zA-Z\ _-]+)\".*" "\\1"
+        ENSMALLEN_VERSION_NAME "${_ensmallen_HEADER_CONTENTS}")
+
+  endif()
+
+  set(ENSMALLEN_VERSION_STRING "${ENSMALLEN_VERSION_MAJOR}.${ENSMALLEN_VERSION_MINOR}.${ENSMALLEN_VERSION_PATCH}")
+endif ()
+
+mark_as_advanced(ENSMALLEN_INCLUDE_DIR)
+
+if (ENSMALLEN_FOUND)
+    set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} "${ENSMALLEN_INCLUDE_DIR}")
+else()
+  message(FATAL_ERROR 
+    "Ensmallen library is not found, please make sure to install ensmallen,"
+    " or use the autodownloader. mlpack can not be installed without cereal")
+endif()
+
+if (USE_OPENMP)
+  find_package(OpenMP)
+endif ()
+
+if (OpenMP_FOUND AND OpenMP_CXX_VERSION VERSION_GREATER_EQUAL 3.0.0)
+  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
+  set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${OpenMP_CXX_LIBRARIES})
+else ()
+  # Disable warnings for all the unknown OpenMP pragmas.
+  if (NOT MSVC)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unknown-pragmas")
+  else ()
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /wd4068")
+  endif ()
+  set(OpenMP_CXX_FLAGS "")
+endif ()
```

Comment:
OpenMP is an optional dependency, so maybe we should add an `MLPACK_DISABLE_OPENMP` so that users can do

```
set(MLPACK_DISABLE_OPENMP 1)
find_package(mlpack ...)
```

if they don't want OpenMP to be enabled?

Reply:
Yeah, you can do that to configure mlpack directly to not use OpenMP, but if a user has copied `Findmlpack.cmake` to their project, then the `USE_OPENMP` option doesn't exist (that is defined in mlpack's `CMakeLists.txt`).  So allowing a way for `Findmlpack.cmake` to disable OpenMP could be useful.  (And we could in turn make our `USE_OPENMP` option in `CMakeLists.txt` simply set `MLPACK_DISABLE_OPENMP` if needed before calling `find_package(mlpack ...)`.)